GameServerLib/ObjectManager.cs

```diff
@@ -134,16 +139,17 @@ public void Update(float diff)
                 var ai = u as IObjAiBase;
                 if (ai != null)
                 {
-                    var tempBuffs = ai.GetBuffs();
-                    for (int i = tempBuffs.Count - 1; i >= 0; i--)
+                    var buffQueue = ai.Buffs.GetQueue();  
+                    while(buffQueue.Count > 0)
```

Comment:
If you already iterate all the buffs to update them, there is no point in using the queue, as you spend greater computational resources on the building of the queue (O(n)/O(nlog n) depends on implementation) and on the constant dequeue (which is usually O(log(n))). Ultimately, this will take much more time than O(n) checks.

The reason for a queue is to identify buffs that needs to be removed faster in case this will be needed, and to handle the priority in sort of pipeline (so FIFO ordering which also allows us to know the elapsed times in order, as it is crucial for some case uses).

This current loops is very inefficient.

Reply:
Just to add some more info. Many buffs don't even need `udate` and their `OnUpdate` is simply empty. We should at some point (not now) keep track on these types of buffs by probably storing them in a separate queue.

Reply:
You mean like this ?

```cs
private SimplePriorityQueue<IBuff> _buffQueue;

public void Update(float diff)
{
    // remove elapsed buffs
    while (_buffQueue.First.TimeElapsed + diff >= _buffQueue.First.Duration)
        _buffQueue.Dequeue();

    // update remaining buffs
    foreach (var buff in _buffQueue)
    {
        buff.Update(diff);
        _buffQueue.UpdatePriority(buff, buff.Duration - buff.TimeElapsed);
    }
}
```